### PR TITLE
Improve the watch API performance and correctness for postgres

### PR DESCRIPTION
**Pull request by jakedt**

### CODE REVIEW

File: pkg/cmd/datastore/datastore.go [hunk 1]
```diff
@@ -123,6 +123,7 @@ func RegisterDatastoreFlags(cmd *cobra.Command, opts *Config) {
 	cmd.Flags().StringVar(&opts.SpannerEmulatorHost, "datastore-spanner-emulator-host", "", "URI of spanner emulator instance used for development and testing (e.g. localhost:9010)")
 	cmd.Flags().StringVar(&opts.TablePrefix, "datastore-mysql-table-prefix", "", "prefix to add to the name of all SpiceDB database tables")
 	cmd.Flags().StringVar(&opts.MigrationPhase, "datastore-migration-phase", "", "datastore-specific flag that should be used to signal to a datastore which phase of a multi-step migration it is in")
+	cmd.Flags().Uint16Var(&opts.WatchBufferLength, "datastore-watch-buffer-length", 128, "how many events the watch buffer should queue before forcefully disconnecting reader")
```
Comment (josephschorr) [inline pkg/cmd/datastore/datastore.go:126]: `128` seems pretty small; maybe a slightly larger default like `1024`?

File: internal/datastore/common/changes.go [hunk 2]
```diff
@@ -11,49 +10,43 @@ import (
...
-		panic(fmt.Errorf("unparseable revision key(%s): %w", key, err))
-	}
-	return rev
-}
-
 // Changes represents a set of tuple mutations that are kept self-consistent
 // across one or more transaction revisions.
-type Changes map[revisionKey]*changeRecord
+type Changes[R datastore.Revision, K comparable] struct {
+	records map[K]*changeRecord[R]
```
Comment (josephschorr) [inline internal/datastore/common/changes.go:16]: why a pointer to changeRecord here?

File: internal/datastore/common/changes.go [hunk 3]
```diff
@@ -75,47 +68,40 @@ func (ch Changes) AddChange(
 	}
 }
 
+type RevisionChanges[K datastore.Revision] struct {
```
Comment (josephschorr) [inline internal/datastore/common/changes.go:71]: doc comment on exported type

File: internal/datastore/mysql/watch.go [hunk 4]
```diff
@@ -112,7 +114,9 @@ func (mds *Datastore) loadChanges(
 	}
 	defer common.LogOnError(ctx, rows.Close)
 
-	stagedChanges := common.NewChanges()
+	stagedChanges := common.NewChanges(func(r revision.Decimal) int64 {
```
Comment (josephschorr) [inline internal/datastore/mysql/watch.go:117]: maybe we should have this function for comparison defined somewhere once? that way, its not getting redefined per call

File: internal/datastore/mysql/watch.go [hunk 5]
```diff
@@ -156,7 +160,9 @@ func (mds *Datastore) loadChanges(
 		return
 	}
 
-	changes = stagedChanges.AsRevisionChanges(mds)
+	changes = stagedChanges.AsRevisionChanges(func(lhs, rhs int64) bool {
```
Comment (josephschorr) [inline internal/datastore/mysql/watch.go:163]: ditto

File: internal/datastore/postgres/watch.go [hunk 6]
```diff
@@ -86,17 +86,19 @@ func (pgd *pgDatastore) Watch(
...
-					errs <- datastore.NewWatchDisconnectedErr()
-					return
-				}
+				for _, changeToWrite := range changesToWrite {
+					changeToWrite := changeToWrite
 
-				currentTxn = revision
-			}
+					select {
+					case updates <- &changeToWrite:
```
Comment (josephschorr) [inline internal/datastore/postgres/watch.go:94]: Add a comment below here saying "Nothing else to do"; while `select` doesn't allow fallthrough, it felt that way, so better to make it explicit IMO

File: internal/datastore/postgres/watch.go [hunk 7]
```diff
@@ -139,10 +141,30 @@ func (pgd *pgDatastore) getNewRevisions(
 	return ids, nil
 }
 
-func (pgd *pgDatastore) loadChanges(ctx context.Context, revision xid8) (*datastore.RevisionChanges, error) {
+func (pgd *pgDatastore) loadChanges(ctx context.Context, revisions []postgresRevision) ([]datastore.RevisionChanges, error) {
+	min := revisions[0].tx.Uint
```
Comment (josephschorr) [inline internal/datastore/postgres/watch.go:152]: is there any chance the revisions slice is empty?

File: internal/datastore/postgres/watch.go [hunk 8]
```diff
@@ -189,21 +213,19 @@ func (pgd *pgDatastore) loadChanges(ctx context.Context, revision xid8) (*datast
...
 		}
 
-		if createdXID.Uint == revision.Uint {
-			tracked.AddChange(ctx, postgresRevision{revision, noXmin}, nextTuple, core.RelationTupleUpdate_TOUCH)
-		} else if deletedXID.Uint == revision.Uint {
-			tracked.AddChange(ctx, postgresRevision{revision, noXmin}, nextTuple, core.RelationTupleUpdate_DELETE)
+		if _, found := filter[createdXID.Uint]; found {
+			tracked.AddChange(ctx, postgresRevision{createdXID, noXmin}, nextTuple, core.RelationTupleUpdate_TOUCH)
+		}
+		if _, found := filter[deletedXID.Uint]; found {
```
Comment (josephschorr) [inline internal/datastore/postgres/watch.go:226]: is it possible for both these conditions to be true?

File: internal/datastore/common/changes.go (hunk 2 above)
Comment (jakedt) [reply internal/datastore/common/changes.go:16]: No good reason, changed it.

File: internal/datastore/postgres/watch.go (hunk 7 above)
Comment (jakedt) [reply internal/datastore/postgres/watch.go:152]: Not the way the code is currently written. It's literally under a test for that exact condition:

```
			if len(newTxns) > 0 {
				changesToWrite, err := pgd.loadChanges(ctx, newTxns)
```

File: internal/datastore/postgres/watch.go (hunk 8 above)
Comment (jakedt) [reply internal/datastore/postgres/watch.go:226]: Yes, if the revision range is large enough to encompass the tuple being both created and deleted.

**Comment by jakedt** [comment]
Updated

File: internal/datastore/mysql/watch.go (hunk 5 above)
Comment (josephschorr) [reply internal/datastore/mysql/watch.go:161]: was this changed?

File: internal/datastore/mysql/watch.go (hunk 4 above)
Comment (jakedt) [reply internal/datastore/mysql/watch.go]: https://github.com/authzed/spicedb/pull/1039/commits/9f80ab3d645a81f9b49c9a7331c61dac48833172#diff-63655d1aac45a55ee99ed6cc29a10501150a30470bc2c600bc3b6355fd315f9eL117

**Review by josephschorr — APPROVED** [review]
LGTM

**Pull request merged**